### PR TITLE
Nix build for midnight and some nix improvements

### PR DESCRIPTION
**Pull request by teodanciu**

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -1,76 +1,79 @@
...
+let
+  customAgda = (import (builtins.fetchTarball
     https://github.com/nixos/nixpkgs/tarball/a7ecde854aee5c4c7cd6177f54a99d2c1ff28a31) ({ # 21.11
       overlays = [ (import (fetchFromGitHub {
         repo = "agda";
         owner = "input-output-hk";
         rev = "d5ea03b96328f38741efef4535197076ff0e05d5";
         sha256 = "WiadZWZvPWcR49JkiXPiMKW3plRjBlR94wyg/aoEoG8=";
       })).overlay ];
-    } // (if pkgs.system == "aarch64-darwin" then { system = "x86_64-darwin"; } else {}))); let
```
Comment (teodanciu) [inline default.nix:11]: I removed the nested `let` - it seemed more difficult to follow and require a few passing on the variables between the scopes

File: src/MidnightExample/Main.hs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-module MidnightExample.Main (module MidnightExample.Main, module MAlonzo.Code.MidnightExample.HSLedger) where
+module Main (module Main, module MAlonzo.Code.MidnightExample.HSLedger) where
```
Comment (teodanciu) [inline src/MidnightExample/Main.hs:1]: I removed the `MidnightExample` prefix from the module name in order to make it consistent with the main module from `Ledger` and make it easier to extract this in a function. 
We could also go the other way, and change `HSLedgerTest` module to `Ledger.HSLedgerTest` - I just chose one option arbitrarily, but happy to do the other one too

File: default.nix (hunk 1 above)
Comment (WhatisRT) [reply default.nix:11]: Your version is definitely clearer than mine!

File: x.nix [hunk 3]
```diff
@@ -0,0 +1,151 @@
+{ nixpkgs ? import <nixpkgs> {} }:
```
Comment (WhatisRT) [inline x.nix:1]: Is this file a leftover?

File: default.nix [hunk 4]
```diff
@@ -1,116 +1,150 @@
...
+    dir = "MidnightExample";
+    agdaLedgerPath = "MidnightExample/HSLedger.agda";
+    hsMainPath = "MidnightExample/Main.hs";
+  };
+
+  ledgerExecutableSpecMidnight = haskellPackages.callCabal2nix "Agda-ledger-executable-spec-midnight" "${ledgerHsSrcMidnight}" { };
+
+  docs = { dir }:
+    stdenv.mkDerivation {
+      pname = "Simple-utxo-ledger-${dir}-docs";
```
Comment (WhatisRT) [inline default.nix:115]: That name is pretty old, I think this would be better
```suggestion
      pname = "${dir}-docs";
```

File: default.nix [hunk 5]
```diff
@@ -1,116 +1,150 @@
...
+        mkdir -p $out
+        agda -c --ghc-dont-call-ghc --compile-dir $out ${agdaLedgerPath}
+        cp ${hsMainPath} $out
+        cp ${dir}/agda-ledger-executable-spec.cabal $out
+        # Append all the modules generated by MAlonzo to the cabal file
+        find $out/MAlonzo -name "*.hs" -print | sed "s#^$out/#        #;s#\.hs##;s#/#.#g" >> $out/agda-ledger-executable-spec.cabal
+      '';
+    };
+
+  ledgerHsSrc = hsSrc {
```
Comment (WhatisRT) [inline default.nix:97]: There's a bunch of repetition here. What do you think about making a single function that accepts this set and generates all three derivations at once?

Comment (teodanciu) [reply default.nix:97]: Indeed! Let's see how that looks like, trying it now

File: default.nix [hunk 6]
```diff
@@ -1,116 +1,148 @@
...
-  agdaWithPkgs = agdaDeps.agda;
 
-  ledger-hs-src = stdenv.mkDerivation {
-    pname = "Agda-ledger-hs-src";
+  deps = [ agdaStdlib agdaFinset agdaStdlibMeta ];
+  agdaWithPkgs = customAgda.agda.withPackages { pkgs = deps; ghc = pkgs.ghc; };
+
+in
+rec {
+
```
Comment (WhatisRT) [inline default.nix:71]: We need to make agda itself available as well
```suggestion
  agda = agdaWithPkgs;
```

**Pull request merged**